### ISSUE TITLE
[Neo4j on Powershell] - Installation fails when JAVA_HOME env variable not set

### ISSUE THREAD

**Issue by glindroth** [neo4j/neo4j#5326]
After unzipping the neo4j installation for windows the following command in power shell console return java -help output:

'C:\Neo4j' | Start-Neo4jServer -Console -Wait

See attached screen shot for output
![screen shot 2015-09-10 at 13 18 15](https://cloud.githubusercontent.com/assets/12694230/9787971/2ee65aa8-57c7-11e5-935f-28aa9f51b58c.png)

**Comment by glindroth** [comment]
@glennsarti I created this for your consideration.

**Comment by glennsarti** [comment]
@glindroth I'll have a look at it as soon as I can. In the mean time can you send through as much info as possible to repro the problem

**Comment by glindroth** [comment]
OK there are some differences between versions of Windows:

---

Windows 7:
Steps:
1. Follow current manual as specified on http://neo4j.com/docs/snapshot/powershell.html until just before the following command: "'C:\Neo4j' | Start-Neo4jServer -Console -Wait"
2. Install Java (downloaded from here: https://java.com/en/download/)
3. Run "'C:\Neo4j' | Start-Neo4jServer -Console -Wait"

Expected result:
Neo4j should start

Actual Result:
Attached screenshot shows that java help text is displayed

Workaround:
set JAVA_HOME environment variable to installation directory of java

---

Windows 8:
Steps:
1. Follow current manual as specified on http://neo4j.com/docs/snapshot/powershell.html until just before the following command: "'C:\Neo4j' | Start-Neo4jServer -Console -Wait"
2. Install Java (downloaded from here: https://java.com/en/download/)
3. Run "'C:\Neo4j' | Start-Neo4jServer -Console -Wait"

WORKS AS EXPECTED

---

Windows 10:
Steps:
1. Follow current manual as specified on http://neo4j.com/docs/snapshot/powershell.html until just before the following command: "'C:\Neo4j' | Start-Neo4jServer -Console -Wait"
2. Install Java (downloaded from here: https://java.com/en/download/)
3. Run "'C:\Neo4j' | Start-Neo4jServer -Console -Wait"

Expected result:
Neo4j should start

Actual Result:
Attached screenshot shows that java help text is displayed

Workaround:
NONE. Even if we set JAVA_HOME the actual result described occurs.

**Comment by glennsarti** [comment]
@glindroth 
Hi Gustav.  I've had some time to try and replicate the issue and so far I haven't.
1. Can you please try adding `-Verbose` at the end of Start-Neo4jServer so that it shows where java is being executed from
2. Can you try your tests by checking that it fails as expected _before_ you install Java
3. Can you post your java versions and the location of all java.exe files
   There can be some weird issues if it uses c:\windows\system32\java.exe

---

I've noticed a typo in the source code which you or I can raise a separate PR for.  In the mean time you can change `bin\Neo4j-Management\Start-Neo4jServer.ps1` 

From
`if ($PSCmdlet.ShouldProcess("$($JavaCMD.java) $($ShellArgs)", 'Start Neo4j'))`

To
`if ($PSCmdlet.ShouldProcess("$($JavaCMD.java) $($JavaCMD.args)", 'Start Neo4j'))`

https://github.com/neo4j/neo4j/blob/2.3/packaging/standalone/src/main/distribution/shell-scripts/bin/Neo4j-Management/Start-Neo4jServer.ps1#L119

---

**My Test process:**
Windows 7 64bit Ultimate 
- Base install from CDROM
- Apply latest patches
- Turn off firewall

Windows 10 64bit Pro Build 10240 (RTM)
- Base install from CDROM
- Turn off firewall

I'm using virtual machines and snapshots/checkpoints so I can revert to a clean VM before each test.

Test process
- Extract the contents of the Community 2.3.0-M03 Zip file to C:\Neo4j
- Run Powershell as an administrator
- Change script execution policy to bypass
- Import the Neo4j-Management module (import-module ......)
- Execute 'C:\Neo4j' | Start-Neo4jServer -Console -Wait -Verbose    (I've added -Verbose to show more detail in the console)

The execution should fail with 'Unable to locate Java' as Java is not installed
- Keep the Powershell window open
- Install Java (1.8 u60)
- Execute 'C:\Neo4j' | Start-Neo4jServer -Console -Wait -Verbose
  Neo4j should start and with the correct Java installation location

Test scenarios;

Win7
PS 64bit - Java 64bit offline installer (Ok)
PS 64bit - Java 32bit offline installer (Ok)
PS 64bit - Java online installer (Ok)

Win10
PS 64bit - Java 64bit offline installer (Ok)
PS 32bit - Java 64bit offline installer.  Error could not find Java. Expected.

**Comment by glindroth** [comment]
Hi Glenn,
Ran with verbose and got the following output:

---

PS C:\Windows\system32> cd c:
PS C:\Windows\system32> Import-Module C:\neo4j\neo4j-community-2.3-SNAPSHOT\bin\Neo4j-Management.psd1
PS C:\Windows\system32> 'C:\neo4j\neo4j-community-2.3-SNAPSHOT\' | Start-Neo4jServer -Console -Wait -Verbose
VERBOSE: Performing the operation "Start Neo4j" on target "C:\Program Files (x86)\Java\jre1.8.0_60\bin\java.exe ".
Usage: java [-options] class [args...](to execute a class)
   or  java [-options] -jar jarfile [args...](to execute a jar file)
where options include:
    -d32          use a 32-bit data model if available
    -d64          use a 64-bit data model if available
    -client       to select the "client" VM
    -server       to select the "server" VM
                  The default VM is client.

```
-cp <class search path of directories and zip/jar files>
-classpath <class search path of directories and zip/jar files>
              A ; separated list of directories, JAR archives,
              and ZIP archives to search for class files.
-D<name>=<value>
              set a system property
-verbose:[class|gc|jni]
              enable verbose output
-version      print product version and exit
-version:<value>
              Warning: this feature is deprecated and will be removed
              in a future release.
              require the specified version to run
-showversion  print product version and continue
-jre-restrict-search | -no-jre-restrict-search
              Warning: this feature is deprecated and will be removed
              in a future release.
              include/exclude user private JREs in the version search
-? -help      print this help message
-X            print help on non-standard options
-ea[:<packagename>...|:<classname>]
-enableassertions[:<packagename>...|:<classname>]
              enable assertions with specified granularity
-da[:<packagename>...|:<classname>]
-disableassertions[:<packagename>...|:<classname>]
              disable assertions with specified granularity
-esa | -enablesystemassertions
              enable system assertions
-dsa | -disablesystemassertions
              disable system assertions
-agentlib:<libname>[=<options>]
              load native agent library <libname>, e.g. -agentlib:hprof
              see also, -agentlib:jdwp=help and -agentlib:hprof=help
-agentpath:<pathname>[=<options>]
              load native agent library by full pathname
-javaagent:<jarpath>[=<options>]
              load Java programming language agent, see java.lang.instrument
-splash:<imagepath>
              show splash screen with specified image
```

See http://www.oracle.com/technetwork/java/javase/documentation/index.html for more details.
1
PS C:\Windows\system32>

---

In what order do you want me to test further? Do you want me to try it with the change you propose in bin\Neo4j-Management\Start-Neo4jServer.ps1? I am guessing that change resolves the Windows 10 issue?

**Comment by glindroth** [comment]
I also confirmed that running without java installed gave the following output:

---

PS C:> 'C:\neo4j\neo4j-community-2.3-SNAPSHOT\' | Start-Neo4jServer -Console -Wait -Verbose
VERBOSE: Unable to determine Java version
Get-Java : Unable to determine the path to java.exe
At C:\neo4j\neo4j-community-2.3-SNAPSHOT\bin\Neo4j-Management\Start-Neo4jServer.ps1:111 char:18
-       $JavaCMD = Get-Java -Neo4jServer $thisServer -ForServer
-                  ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
  - CategoryInfo          : NotSpecified: (:) [Write-Error], WriteErrorException
  - FullyQualifiedErrorId : Microsoft.PowerShell.Commands.WriteErrorException,Get-Java

Start-Neo4jServer : Unable to locate Java
At line:1 char:44
- ... -community-2.3-SNAPSHOT\' | Start-Neo4jServer -Console -Wait -Verbose
-                                 ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
  - CategoryInfo          : NotSpecified: (:) [Write-Error], WriteErrorException
  - FullyQualifiedErrorId : Microsoft.PowerShell.Commands.WriteErrorException,Start-Neo4jServer

---

**Comment by glindroth** [comment]
I also tested the scenario you described on Win10 after updating the bin\Neo4j-Management\Start-Neo4jServer.ps1 file:

---

Test process

Extract the contents of the Community 2.3.0-M03 Zip file to C:\Neo4j
Run Powershell as an administrator
Change script execution policy to bypass
Import the Neo4j-Management module (import-module ......)
Execute 'C:\Neo4j' | Start-Neo4jServer -Console -Wait -Verbose (I've added -Verbose to show more detail in the console)
The execution should fail with 'Unable to locate Java' as Java is not installed

Keep the Powershell window open
Install Java (1.8 u60)
Execute 'C:\Neo4j' | Start-Neo4jServer -Console -Wait -Verbose Neo4j should start and with the correct Java installation location

---

without success and the following output:

---

PS C:> 'C:\neo4j\neo4j-community-2.3-SNAPSHOT\' | Start-Neo4jServer -Console -Wait -Verbose
VERBOSE: Performing the operation "Start Neo4j" on target "C:\Program Files (x86)\Java\jre1.8.0_60\bin\java.exe
-DworkingDir="C:\neo4j\neo4j-community-2.3-SNAPSHOT\"
-Djava.util.logging.config.file="C:\neo4j\neo4j-community-2.3-SNAPSHOT\conf\windows-wrapper-logging.properties"
-DconfigFile="conf/neo4j-wrapper.conf" -DserverClasspath="lib/_.jar;system/lib/_.jar;plugins/*_/_.jar;./conf*"
-DserverMainClass=org.neo4j.server.CommunityBootstrapper -jar
"C:\neo4j\neo4j-community-2.3-SNAPSHOT\bin\windows-service-wrapper-5.jar"".
Usage: java [-options] class [args...](to execute a class)
   or  java [-options] -jar jarfile [args...](to execute a jar file)
where options include:
    -d32          use a 32-bit data model if available
    -d64          use a 64-bit data model if available
    -client       to select the "client" VM
    -server       to select the "server" VM
                  The default VM is client.

```
-cp <class search path of directories and zip/jar files>
-classpath <class search path of directories and zip/jar files>
              A ; separated list of directories, JAR archives,
              and ZIP archives to search for class files.
-D<name>=<value>
              set a system property
-verbose:[class|gc|jni]
              enable verbose output
-version      print product version and exit
-version:<value>
              Warning: this feature is deprecated and will be removed
              in a future release.
              require the specified version to run
-showversion  print product version and continue
-jre-restrict-search | -no-jre-restrict-search
              Warning: this feature is deprecated and will be removed
              in a future release.
              include/exclude user private JREs in the version search
-? -help      print this help message
-X            print help on non-standard options
-ea[:<packagename>...|:<classname>]
-enableassertions[:<packagename>...|:<classname>]
              enable assertions with specified granularity
-da[:<packagename>...|:<classname>]
-disableassertions[:<packagename>...|:<classname>]
              disable assertions with specified granularity
-esa | -enablesystemassertions
              enable system assertions
-dsa | -disablesystemassertions
              disable system assertions
-agentlib:<libname>[=<options>]
              load native agent library <libname>, e.g. -agentlib:hprof
              see also, -agentlib:jdwp=help and -agentlib:hprof=help
-agentpath:<pathname>[=<options>]
              load native agent library by full pathname
-javaagent:<jarpath>[=<options>]
              load Java programming language agent, see java.lang.instrument
-splash:<imagepath>
              show splash screen with specified image
```

See http://www.oracle.com/technetwork/java/javase/documentation/index.html for more details.
1
PS C:>

---

**Comment by glennsarti** [comment]
@glindroth Thanks Hi Gustav.  That Win10 example was what I was after.  Now I can see the exact command line used to start Java and see what it's complaining about.  I'll start repro-ing it now.

**Comment by glennsarti** [comment]
@glindroth Hi Gustav.  Sorry to be a pain again.  Can you send the -Verbose output on Win10 again.  Unfortunately GitHub has interpreted various characters as markdown.  I'm only really interested in the VERBOSE: output and put the whole thing in a code markdown block e.g.

```
... Text ...
... Text ...
... Text ...
```

**Comment by glennsarti** [comment]
@glindroth  Okay I think I found the problem.  It's the trailing slash on the Pathname.  Instead of `'C:\neo4j\neo4j-community-2.3-SNAPSHOT\'` can you try `'C:\neo4j\neo4j-community-2.3-SNAPSHOT'`

UPDATE: Yup I can repro the issue.  I'll make some changes, write some tests and send up a PR.  Thanks Gustav!

**Comment by glindroth** [comment]
Me too! Verified that it works without the trailing slash. Thanks!

**Comment by glennsarti** [comment]
Raised PR #5446 with fix and tests

**Comment by glindroth** [comment]
Thanks, we will test it again once it's in.

**Comment by spacecowboy** [comment]
Fixed in #5446
